### PR TITLE
Support mop with npm 3+

### PR DESCRIPTION
**Pull request by cdebost**

### CODE REVIEW

**Comment by hthetiot** [comment]
Check this:
- https://github.com/montagejs/mop/blob/3180aba5d511ea8af0817a50ad5ac7c5fdf0cfe1/lib/transform/javascript.js#L32
- https://github.com/montagejs/mr/pull/122

I will check wip

File: montage.js [hunk 1]
```diff
@@ -275,32 +276,32 @@
...
-                    callback(mrRequire, mrPromise, miniURL);
+                    callback(mrRequire, Promise, miniURL);
                 }
 
                 function bootstrapModuleScript(module, err, script) {
                     if (err) {
                         // Fallback on flat strategy for missing nested module
                         if (module.strategy === 'nested') {
                             module.strategy = 'flat';
-                            module.script = resolveUrl(resolveUrl(location, '../../'), module.location);
```
Comment (hthetiot) [inline montage.js]: ok

File: montage.js [hunk 2]
```diff
@@ -323,26 +353,28 @@
...
+                            module.location = params[paramModuleLocation];
+                        }
 
                         // Reset bad exports
                         if (moduleHasExport(module)) {
                             bootstrapModule(module.id, module.exports);
                         } else if (typeof module.shim !== "undefined") {
                             bootstrapModule(module.id, module.shim);
-                        } else {
+                        } else if (!global.BUNDLE) {
```
Comment (hthetiot) [inline montage.js]: ok

File: montage.js [hunk 3]
```diff
@@ -323,26 +353,28 @@
...
-                            module.id = id;   
+                            module.id = id;
                         }
 
                         // Update dependency
-                        dependencies[id] = module;  
-                        // Update locatiom from param
-                        module.location = params.hasOwnProperty(paramModuleLocation) ? params[paramModuleLocation] : module.location;
+                        dependencies[id] = module;
+                        // Update location of dependency from params, e.g. if we are in a mop build
```
Comment (hthetiot) [inline montage.js]: ok

**Comment by hthetiot** [comment]
ok fix the defer that can be factored otherwise look ok.

**Comment by cdebost** [comment]
@hthetiot What do you mean by "defer that can be factored"?

**Comment by hthetiot** [comment]
@cdebost 2 places in the PR you have " = new Promise(" that can be optimise by a simple defer() function I added in comment.

**Comment by hthetiot** [comment]
@cdebost Can you re PR from commonjs-mop-npm3 not form your cdebost like that i can patch instead of fork and re-patch

**Pull request merged**